Add sharding groups to api-tests to allow running in parallel

Smackteo opened the pull request:
#### 3cecfd095c0d1ceed9d0b74a98c0a675d3d49217
<pre>
Add sharding groups to api-tests to allow running in parallel
<a href="https://bugs.webkit.org/show_bug.cgi?id=298892">https://bugs.webkit.org/show_bug.cgi?id=298892</a>
<a href="https://rdar.apple.com/160635460">rdar://160635460</a>

Reviewed by Jonathan Bedard.

This adds a mutually exclusive shard group called system that is used when api-tests are run.
Anything in the System will not be run in parallel for api-tests, this also changes the default --child-processes argument in run-api-tests to allow parallelization to be the new default.

* Tools/Scripts/webkitpy/api_tests/runner.py:
(Runner.run):
* Tools/Scripts/webkitpy/port/base.py:
(Port.sharding_groups):
(Port.group_for_shard):
* Tools/Scripts/webkitpy/port/darwin.py:
(DarwinPort.sharding_groups):

Canonical link: <a href="https://commits.webkit.org/300034@main">https://commits.webkit.org/300034@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a90577f44cf9e67ee4957edc5de1164a28a99b7d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121122 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/40818 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31476 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127541 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73203 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/15956c52-adb4-42ef-803b-caee2017135a) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/122998 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41520 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49397 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/92008 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/61213 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/479c5953-5044-46ad-a997-2927fa55b21e) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124074 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/33163 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/108574 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72691 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/220f3448-09d3-45b5-a3ed-42b46655804e) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/120479 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/32187 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26677 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71133 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/102667 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/26856 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130394 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48049 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36520 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/100616 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48417 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104741 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100519 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25478 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/45924 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/23974 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/44743 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/47907 "Built successfully") | | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47378 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/50725 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49062 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/Scripts/webkitpy/api_tests/runner.py:
```diff
@@ -146,10 +146,25 @@ def run(self, tests, num_workers):
 
             with TaskPool(
                 workers=self._num_workers,
+                mutually_exclusive_groups=list(self.port.sharding_groups(suite='api-tests').keys()),
```
You are going to want to save this value and operate differently if it's not defined (that's particularly relevant for non-Apple platforms)

JonWBedard commented on Tools/Scripts/webkitpy/api_tests/runner.py:
```diff
@@ -146,10 +146,25 @@ def run(self, tests, num_workers):
 
             with TaskPool(
                 workers=self._num_workers,
```
If `mutually_exclusive_groups` is defined, you want `_num_workers` to be half the number of cores by default, I think. Maybe the same as the number of cores.

Smackteo replied:
By absolute default `Tools/Scripts/webkitpy/port/base.py ` uses the same as number of cores. 
If I remove the default of `'--child-processes=1'` from `Tools/Scripts/run-api-tests` it will run in that fashion.

JonWBedard commented on Tools/Scripts/webkitpy/api_tests/runner.py:
```diff
@@ -135,7 +135,11 @@ def run(self, tests, num_workers):
                 raise RuntimeError('Cannot nest API test runners')
             Runner.instance = self
-            self._num_workers = min(num_workers, len(shards))
+            mutually_exclusive_groups = list(self.port.sharding_groups(suite='api-tests').keys())
```
I'm surprised to see us converting this to a list. The dictionary you're returning contains the callbacks that `TaskPool` needs, does it not?

Smackteo replied:
This is how it's done in `layout_test_runner.py line 168:`
`mutually_exclusive_groups=list(self._port.sharding_groups().keys()),`

TaskPool's mutually_exclusive_groups parameter expects a list of group names, not the callback functions.
The callbacks are used by group_for_shard()

JonWBedard commented on Tools/Scripts/webkitpy/port/darwin.py:
```diff
@@ -53,7 +53,11 @@ def __init__(self, host, port_name, **kwargs):
+    def sharding_groups(self, suite=None):
+        if suite == 'api-tests':
+            return {
+                'system': lambda shard: shard.name.startswith('TestWebKitAPI') or shard.name.startswith('TestWGSL'),
```
You'll need something more substantive than a lambda soon, but this is fine for now.

Smackteo committed:
Add sharding groups to api-tests to allow running in parallel
https://bugs.webkit.org/show_bug.cgi?id=298892
rdar://160635460

Reviewed by Jonathan Bedard.

This adds a mutually exclusive shard group called system that is used when api-tests are run.
Anything in the System will not be run in parallel for api-tests, this also changes the default --child-processes argument in run-api-tests to allow parallelization to be the new default.

* Tools/Scripts/webkitpy/api_tests/runner.py:
(Runner.run):
* Tools/Scripts/webkitpy/port/base.py:
(Port.sharding_groups):
(Port.group_for_shard):
* Tools/Scripts/webkitpy/port/darwin.py:
(DarwinPort.sharding_groups):

Canonical link: https://commits.webkit.org/300034@main
